src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs

```diff
@@ -0,0 +1,148 @@
+                }
+            }
+
+            return null;
+        }
+
+        private static async Task<bool> CheckTypeInfoOfAttributeAsync(CompletionContext context, SyntaxNode attributeNode)
+        {
+            var semanticModel = await context.Document.GetSemanticModelForNodeAsync(attributeNode, context.CancellationToken).ConfigureAwait(false);
+            var typeInfo = semanticModel.GetTypeInfo(attributeNode);
+            var type = typeInfo.Type;
+            if (type == null)
+            {
+                return false;
+            }
+
+            //TODO there must be better ways to do the check type equality.
+            var typeName = type.ToDisplayString(new SymbolDisplayFormat(typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces));
+            var assemblyName = type.ContainingAssembly.ToSignatureDisplayString();
+            var internalsVisibleToQualifiedName = typeof(InternalsVisibleToAttribute).AssemblyQualifiedName;
+            return typeName + ", " + assemblyName == internalsVisibleToQualifiedName;
+        }
+
+        private static void AddAssemblyCompletionItems(CompletionContext context, CancellationToken cancellationToken)
+        {
+            var currentProject = context.Document.Project;
+            foreach (var project in context.Document.Project.Solution.Projects)
+            {
+                if (project == currentProject)
+                {
+                    continue;
+                }
+
+                var projectGuid = project.Id.Id.ToString();
+                var completionItem = CommonCompletionItem.Create(
+                    displayText: project.AssemblyName,
+                    rules: CompletionItemRules.Default,
+                    glyph: project.GetGlyph(),
+                    properties: ImmutableDictionary.Create<string, string>().Add(ProjectGuidKey, projectGuid));
+                context.AddItem(completionItem);
+            }
+        }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = default(char?), CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (item.Properties.TryGetValue(ProjectGuidKey, out var projectId))
+            {
+                var assemblyName = item.DisplayText;
+                var project = document.Project.Solution.GetProject(ProjectId.CreateFromSerialized(new System.Guid(projectId)));
+                var publicKey = await GetPublicKeyOfProjectAsync(project, cancellationToken).ConfigureAwait(false);
+                if (!string.IsNullOrEmpty(publicKey))
+                {
+                    assemblyName += ", PublicKey=" + publicKey;
+                }
+                var textChange = new TextChange(item.Span, assemblyName);
+                return CompletionChange.Create(textChange);
+            }
+
+            Debug.Fail("Project can't be found by projectId.");
```

Comment:
should be impossible.  So it should be fine to just index directly into item.Properties.